### PR TITLE
Simplify frontMatter parse

### PR DESCRIPTION
**Pull request by fisker**
## Description

I'm not sure if I made this harder to read.

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/utils/front-matter.js [hunk 1]
```diff
@@ -1,55 +1,34 @@
...
-    .join("|");
-
-  const match = text.match(
-    // trailing spaces after delimiters are allowed
-    new RegExp(
-      `^(${startDelimiterRegex})([^\\n]*)\\n(?:|([\\s\\S]*?)\\n)(${endDelimiterRegex})[^\\n\\S]*(\\n|$)`
-    )
-  );
+// trailing spaces after delimiters are allowed
+const frontMatterRegex = /^(?<startDelimiter>-{3}|\+{3})(?<language>[^\n]*)\n(?:|(?<value>[\S\s]*?)\n)(?<endDelimiter>\k<startDelimiter>|\.{3})[^\S\n]*(?:\n|$)/;
```
Comment (thorn0) [inline src/utils/front-matter.js:11]: I'd rewrite it for readability:
```
const frontMatterRegex = new RegExp([
  "^(?<startDelimiter>-{3}|\\+{3})" +
    "(?<language>[^\\n]*)\\n" +
    "(?:|(?<value>[\\S\\s]*?)\\n)" +
    "(?<endDelimiter>k<startDelimiter>|\\.{3})" +
    "[^\\S\\n]*(?:\\n|$)",
]);
```

**Review by thorn0 — APPROVED** [review]

File: src/utils/front-matter.js (hunk 1 above)
Comment (fisker) [reply src/utils/front-matter.js:11]: I hate double escape regular expressions. Maybe I can use `Regex#source` to make it smore readable?

Comment (thorn0) [reply src/utils/front-matter.js:11]: Nobody likes that escaping, but IMHO it's the lesser evil.

(I wish JS had free-spacing regexps (`x` flag). Prettier would be able to reformat such regexps...)

> Maybe I can use Regex#source to make it more readable?

Can't imagine what you mean.

Comment (fisker) [reply src/utils/front-matter.js:11]:
```js
const frontMatterRegex = new RegExp(
  /^(?<startDelimiter>-{3}|\+{3})/.source +
    /(?<language>[^\n]*)/.source +
    /\n(?:|(?<value>[\S\s]*?)\n)/.source +
    "(?<endDelimiter>\\k<startDelimiter>|\\.{3})" +
    /[^\S\n]*(?:\n|$)/.source
);
```

Comment (fisker) [reply src/utils/front-matter.js:11]: Put "\n" on 3rd part, prevent `unicorn/better-regex` remove `(?:` on that group. 4th part can't write in regex literal "invalid group".

Comment (fisker) [reply src/utils/front-matter.js:11]: This lib looks cool http://xregexp.com/

**Pull request merged**